### ISSUE TITLE
0.14 is broken in IE11

### ISSUE THREAD

**Issue by mattlewis92** [sahat/satellizer#753]
It looks like the changes between 0.13 and 0.14 broke this library in IE11, after signing in ok to the popup window it redirects back to the app and never closes the popup + continues the process of adding the account. Rolling back from 0.14 to 0.13.4 fixes the problem.

**Comment by sahat** [comment]
After looking into it further, seems like IE is not normalizing URL pathname to "/", like other browsers. I will try to release a patch soon that fixes this issue.

**Comment by mattlewis92** [comment]
Fantastic, thanks for the quick response!

**Comment by sahat** [comment]
Try the latest `satellizer.js` file, if that fixes your problem with IE, I will bump the version number and publish it on NPM.

**Comment by mattlewis92** [comment]
I gave your fix a try and it doesn't seem to work, would you like me to try and put together a minimally failing example?

**Comment by sahat** [comment]
Sure, but I just find it strange that didn't work for you. I was able to reproduce the problem in IE11 and pretty certain it is related to URL pathname mismatch.

Do you have Protected Mode turned off in IE options?

**Comment by mattlewis92** [comment]
I don't think I have protected mode enabled, I'm just using the IE VM from the modern.ie site. How can I double check? I'll put together an example tomorrow for you.

**Comment by sahat** [comment]
By default it is turned on. The issue you are describing occurs only when running the app locally. If you are to deploy it first, to Heroku for example, then visit that URL from your IE VM, it should work. If it doesn't, then it is likely another issue.
![sshot-1](https://cloud.githubusercontent.com/assets/544954/13446248/d0b89698-dfdf-11e5-8e83-8564fc5e221c.png)

**Comment by sahat** [comment]
Hmm you did say it works with 0.13.4, so maybe I was wrong about Protected Mode issue.

**Comment by mattlewis92** [comment]
Yup it's deployed on a proper domain with ssl and it's still happening

**Comment by LuisMejiaF** [comment]
I'm experiencing the same issue when trying with authenticate with Google or Facebook: Satellizer opens the popup but it remains open and nothing happens.

I noticed that the URL has the **code** parameter, but the structure is such that angular cannot find the route.

For example: my application redirects on call to **mysite.com/#/**, which is the homepage. But satellizer facebook or google somehow redirect to **mysite.com/?code=xxxxxx/#/**, and in that case angular cannot find the route. 

I'm using Chrome version 48.0.2564.116 in windows desktop.

**Comment by mattlewis92** [comment]
I dug into this a bit more today and have found the cause of the problem.

I put a debug statement [here](https://github.com/sahat/satellizer/blob/master/satellizer.js#L767) with `console.log('DEBUG', popupWindowPath, redirectUriPath);`

<img width="446" alt="screen shot 2016-03-02 at 10 53 18" src="https://cloud.githubusercontent.com/assets/6425649/13458523/a11407c0-e065-11e5-95a2-2f8fc636fa1b.png">

As you can see `popupWindowPath` doesn't have the port number and `redirectUriPath` has the port number so the popup never closes.

**Comment by sahat** [comment]
Ah, yet another IE's weird behavior. Both Chrome and Firefox return empty port number. I am not sure if it's the best way, but I could explicitly ignore port number concatenation if port === "80" or port === "443" (http and https ports, respectively). 

These minor differences, like pathname is empty string in IE and slash in Chrome/Firefox, or empty port number in Chrome/Firefox and port 80 in Internet Explorer stem from this code:

```js
var redirectUriParser = document.createElement('a');
redirectUriParser.href = redirectUri;
```

It is necessary to parse redirectUri so issues like popup not closing wouldn't occur if redirectUri is missing a slash at the end, for example **http://localhost:3000** !== **http://localhost:3000/**.

**Comment by mattlewis92** [comment]
What about going the other way round and defining the port number to 80 or 443 if not set and the protocol is http / https respectively?

**Comment by sahat** [comment]
Is there any advantage to doing so? In either case I will have about the same amount of work: 1) not using a port number if port is 80 or 443, or 2) using port number 80 if http, 443 if https, or use existing port number if available.

**Comment by mattlewis92** [comment]
Normalising the data just seems like a cleaner implementation to me, but either would work fine.

**Comment by sahat** [comment]
Alright, while I am writing unit tests for this function, could you provide me some URLs to test potential edge cases? I want to solve this problem once and for all.

**Comment by mattlewis92** [comment]
Sure thing, just to be sure we're on the same page do you mean [this function](https://github.com/sahat/satellizer/blob/master/satellizer.js#L829)?

As a side note, have you ever tried browserstack before? It's free for open source and you can have travis run all your tests in IE9-11, FF etc to catch browser differences like this in the future. It's super easy to integrate as you're already using karma, just a few extra bits added to your [config](https://github.com/karma-runner/karma-browserstack-launcher)

**Comment by sahat** [comment]
Yes that's the function getFullUrlPath.

I have heard of BrowserStack but did not realize they offer a free tier for open-source projects. I'll check them out. It's a pain testing all browsers with every release.

**Comment by mattlewis92** [comment]
It looks like you have [to apply](https://www.browserstack.com/accounts/subscriptions), but I doubt they'd say no for a project this popular. 

For test cases, extracted from running this in each browsers console:
```
var a = document.createElement('a'); a.href='http://google.com/test'; console.log(a.protocol, a.hostname, a.port, a.pathname);
```

Chrome, Safari, FF, domain `http://google.com/test`
```
{
  protocol: 'http:',
  hostname: 'google.com',
  port: '',
  pathname: '/test'
}
```

Chrome, Safari, FF, domain `https://google.com/test`
```
{
  protocol: 'https:',
  hostname: 'google.com',
  port: '',
  pathname: '/test'
}
```

IE 9-11, domain `http://google.com/test`
```
{
  protocol: 'http:',
  hostname: 'google.com',
  port: '80',
  pathname: 'test'
}
```

IE 9-11, domain `https://google.com/test`
```
{
  protocol: 'https:',
  hostname: 'google.com',
  port: '443',
  pathname: 'test'
}
```

**Comment by sahat** [comment]
Thanks. Alright, try satellizer.js from master branch again. It should normalize the port number now.

**Comment by LuisMejiaF** [comment]
It is fixed for me in Chrome Desktop, but it fails in Mobile browers in Android, I get the same behavior that I got in Chrome earlier.

**Comment by mattlewis92** [comment]
@sahat the fixes in master are working perfectly for me now in IE11, thanks for that!

@LuisMejiaF I checked the output of `var a = document.createElement('a'); a.href='http://google.com/test'; console.log(a.protocol, a.hostname, a.port, a.pathname);` in an android VM and it gave the same output as chrome. It's either a different bug, or a discrepancy in your specific android browser (which are well known for varying wildly in implementations, hence the existence of crosswalk etc)

**Comment by sahat** [comment]
@mattlewis92 Excellent. Thanks for your help too. I will push a new version soon.

**Comment by mattlewis92** [comment]
Awesome, thanks!

**Comment by fatcop** [comment]
any new tagged version for this fix imminent ?

**Comment by barryvdh** [comment]
I've run into this issue as well. Seems to be present in the latest release, but the current dev-master is working. I've installed `bower install "https://github.com/sahat/satellizer.git#ae7912b8e100a0ec86f66526c0b342b5392189b3"` for now, but a tag would be much appreciated!

**Comment by clarle** [comment]
Fixed in 0.14.1.

Let us know if you see any other issues, and thanks for reporting!
